### PR TITLE
Making the API default missing values 

### PR DESCRIPTION
**Pull request by diptanu**

### CODE REVIEW

File: api/jobs.go [hunk 1]
```diff
@@ -207,10 +219,36 @@ type UpdateStrategy struct {
...
+}
+
+func (p *PeriodicConfig) Canonicalize() {
+	if p.Enabled == nil {
+		p.Enabled = helper.BoolToPtr(true)
+	}
+	if p.Spec != nil {
+		p.SpecType = helper.StringToPtr(PeriodicSpecCron)
+	}
+	if p.ProhibitOverlap == nil {
```
Comment (dadgar) [inline api/jobs.go]: SpecType?

File: api/jobs.go [hunk 2]
```diff
@@ -237,12 +275,90 @@ type Job struct {
...
+		j.Region = helper.StringToPtr("global")
+	}
+	if j.Type == nil {
+		j.Type = helper.StringToPtr("service")
+	}
+	if j.Periodic != nil {
+		j.Periodic.Canonicalize()
+	}
+
+	for _, tg := range j.TaskGroups {
```
Comment (dadgar) [inline api/jobs.go]: Break these up into different functions

File: api/jobs.go [hunk 3]
```diff
@@ -237,12 +275,90 @@ type Job struct {
...
+		for _, task := range tg.Tasks {
+			if task.LogConfig == nil {
+				task.LogConfig = &LogConfig{
+					MaxFiles:      helper.IntToPtr(10),
+					MaxFileSizeMB: helper.IntToPtr(10),
+				}
+			} else {
+				if task.LogConfig.MaxFiles == nil {
+					task.LogConfig.MaxFiles = helper.IntToPtr(10)
```
Comment (dadgar) [inline api/jobs.go]: The fields should be handled separately

File: api/jobs.go [hunk 4]
```diff
@@ -237,12 +275,90 @@ type Job struct {
...
+					task.LogConfig.MaxFileSizeMB = helper.IntToPtr(10)
+				}
+			}
+			if task.Vault != nil {
+				if task.Vault.Env == nil {
+					task.Vault.Env = helper.BoolToPtr(true)
+					task.Vault.ChangeMode = helper.StringToPtr("restart")
```
Comment (dadgar) [inline api/jobs.go]: These should be handled separately

Comment (dadgar) [inline api/jobs.go]: Handle changemode being signal but signal not being set?

File: api/jobs_test.go [hunk 5]
```diff
@@ -50,6 +50,19 @@ func TestJobs_Register(t *testing.T) {
...
+	c, s := makeClient(t, nil, nil)
+	defer s.Stop()
+	jobs := c.Jobs()
+
+	// Create a job and attempt to register it
+	job := testJob()
+	_, err := jobs.Validate(job, nil)
+	if err != nil {
+		t.Fatalf("err: %s", err)
+	}
```
Comment (dadgar) [inline api/jobs_test.go]: Add a failure case

File: command/agent/job_endpoint.go [hunk 6]
```diff
@@ -111,6 +114,40 @@ func (s *HTTPServer) jobPlan(resp http.ResponseWriter, req *http.Request,
...
+	job := s.apiJobToStructJob(validateRequest.Job)
+	args := structs.JobValidateRequest{
+		Job: job,
+	}
+	s.parseRegion(req, &args.Region)
+
+	var out structs.JobValidateResponse
+	if err := s.agent.RPC("Job.Validate", &args, &out); err != nil {
+
+		// Fall back to do local validation
```
Comment (dadgar) [inline command/agent/job_endpoint.go]: Set that the driver configs were not validated

File: command/agent/job_endpoint.go [hunk 7]
```diff
@@ -111,6 +114,40 @@ func (s *HTTPServer) jobPlan(resp http.ResponseWriter, req *http.Request,
 	return out, nil
 }
 
+func (s *HTTPServer) ValidateJobRequest(resp http.ResponseWriter, req *http.Request) (interface{}, error) {
+	var validateRequest api.JobValidateRequest
```
Comment (dadgar) [inline command/agent/job_endpoint.go]: Check that the request is PUT/POST

File: command/agent/job_endpoint.go [hunk 8]
```diff
@@ -310,3 +347,86 @@ func (s *HTTPServer) jobDispatchRequest(resp http.ResponseWriter, req *http.Requ
 	setIndex(resp, out.Index)
 	return out, nil
 }
+
+func (s *HTTPServer) apiJobToStructJob(job *api.Job) *structs.Job {
+	if job.Priority == nil {
```
Comment (dadgar) [inline command/agent/job_endpoint.go]: Why is this not api.Canonicalize() and then just setting each field. It is probably also worth adding apiJobToStructJob which calls apiTaskGroupsToStructTaskGroups and so on.

File: command/plan.go [hunk 9]
```diff
@@ -129,12 +112,18 @@ func (c *PlanCommand) Run(args []string) int {
 	}
 
 	// Force the region to be that of the job.
-	if r := job.Region; r != "" {
+	if r := *job.Region; r != "" {
```
Comment (dadgar) [inline command/plan.go]: This can be a nil pointer dereference since the API job has not been canonicalized

File: command/plan.go [hunk 10]
```diff
@@ -129,12 +112,18 @@ func (c *PlanCommand) Run(args []string) int {
...
 
+	// Check that the job is valid
+	if _, err := client.Jobs().Validate(job, nil); err != nil {
+		c.Ui.Error(fmt.Sprintf("Error validating job: %s", err))
+		return 255
+	}
+
 	// Submit the job
-	resp, _, err := client.Jobs().Plan(apiJob, diff, nil)
+	resp, _, err := client.Jobs().Plan(job, diff, nil)
```
Comment (dadgar) [inline command/plan.go]: Can Plan just validate the job and then you don't need two RPCs. I think it already does

File: command/plan.go [hunk 11]
```diff
@@ -192,7 +181,7 @@ func formatDryRun(resp *api.JobPlanResponse, job *structs.Job) string {
 		out = "[bold][green]- All tasks successfully allocated.[reset]\n"
 	} else {
 		// Change the output depending on if we are a system job or not
-		if job.Type == "system" {
+		if *job.Type == "system" {
```
Comment (dadgar) [inline command/plan.go]: This could be a nil pointer too

File: command/run.go [hunk 12]
```diff
@@ -132,50 +133,52 @@ func (c *RunCommand) Run(args []string) int {
...
+	if r := *job.Region; r != "" {
+		client.SetRegion(r)
+	}
+
 	// Check if the job is periodic or is a parameterized job
 	periodic := job.IsPeriodic()
 	paramjob := job.IsParameterized()
 
+	// Check that the job is valid
+	if _, err := client.Jobs().Validate(job, nil); err != nil {
```
Comment (dadgar) [inline command/run.go]: I think we can remove this as well if the register endpoint validates which I am pretty sure it does

File: command/run.go [hunk 13]
```diff
@@ -132,50 +133,52 @@ func (c *RunCommand) Run(args []string) int {
...
+		job.VaultToken = helper.StringToPtr(vaultToken)
 	}
 
 	// COMPAT 0.4.1 -> 0.5 Remove in 0.6
-	if apiJob.TaskGroups != nil {
+	if job.TaskGroups != nil {
 	OUTSIDE:
-		for _, tg := range apiJob.TaskGroups {
+		for _, tg := range job.TaskGroups {
 			if tg.Tasks != nil {
```
Comment (dadgar) [inline command/run.go]: Don't need this if. An iterator over nil slice will work

File: command/run.go [hunk 14]
```diff
@@ -132,50 +133,52 @@ func (c *RunCommand) Run(args []string) int {
...
 				for _, task := range tg.Tasks {
 					if task.Resources != nil {
-						if task.Resources.DiskMB > 0 {
+						if task.Resources.DiskMB != nil {
```
Comment (dadgar) [inline command/run.go]: In the validation response we should add a list of warnings and then we can add this to the list and push this server side

File: command/run.go [hunk 15]
```diff
@@ -240,7 +231,7 @@ func (c *RunCommand) Run(args []string) int {
 	}
 
 	// Check if we should enter monitor mode
-	if detach || periodic || paramjob {
+	if detach || job.Periodic != nil || job.ParameterizedJob != nil {
```
Comment (dadgar) [inline command/run.go]: Use the helpers you added

File: nomad/job_endpoint.go [hunk 16]
```diff
@@ -285,6 +285,25 @@ func (j *Job) Summary(args *structs.JobSummaryRequest,
...
 }
 
+// Validate validates a job
+func (j *Job) Validate(args *structs.JobValidateRequest,
+	reply *structs.JobValidateResponse) error {
+
+	if err := validateJob(args.Job); err != nil {
+		if merr, ok := err.(*multierror.Error); ok {
+			for _, err := range merr.Errors {
+				reply.ValidationErrors = append(reply.ValidationErrors, err.Error())
```
Comment (dadgar) [inline nomad/job_endpoint.go]: I think you want to append the list of errors internal to multierror

File: nomad/structs/structs.go [hunk 17]
```diff
@@ -282,6 +282,18 @@ type JobDispatchRequest struct {
...
 
+// JobValidateRequest is used to validate a job
+type JobValidateRequest struct {
+	Job *Job
+	WriteRequest
+}
+
+// JobValidateResponse is the response from validate request
+type JobValidateResponse struct {
+	DriverConfigValidated bool
```
Comment (dadgar) [inline nomad/structs/structs.go]: Add comments

File: api/jobs.go [hunk 18]
```diff
@@ -371,6 +487,11 @@ func (j *Job) AddPeriodicConfig(cfg *PeriodicConfig) *Job {
 	return j
 }
 
+// JobValidateRequest is used to validate a job
+type JobValidateRequest struct {
+	Job *Job
```
Comment (dadgar) [inline api/jobs.go]: WriteRequest?

File: jobspec/parse.go [hunk 19]
```diff
@@ -196,18 +199,18 @@ func parseJob(result *structs.Job, list *ast.ObjectList) error {
...
-		result.TaskGroups = make([]*structs.TaskGroup, len(tasks), len(tasks)*2)
+		result.TaskGroups = make([]*api.TaskGroup, len(tasks), len(tasks)*2)
 		for i, t := range tasks {
-			result.TaskGroups[i] = &structs.TaskGroup{
+			result.TaskGroups[i] = &api.TaskGroup{
 				Name:          t.Name,
 				Count:         1,
-				EphemeralDisk: structs.DefaultEphemeralDisk(),
-				Tasks:         []*structs.Task{t},
+				EphemeralDisk: &api.EphemeralDisk{SizeMB: 300},
```
Comment (dadgar) [inline jobspec/parse.go]: Those should be pointers ^ (we decided on all of them being pointers)

File: jobspec/parse.go [hunk 20]
```diff
@@ -317,7 +324,7 @@ func parseGroups(result *structs.Job, list *ast.ObjectList) error {
 		}
 
 		// Parse ephemeral disk
-		g.EphemeralDisk = structs.DefaultEphemeralDisk()
+		g.EphemeralDisk = &api.EphemeralDisk{SizeMB: 300}
```
Comment (dadgar) [inline jobspec/parse.go]: Pointer ^

File: jobspec/parse.go [hunk 21]
```diff
@@ -1065,17 +1089,22 @@ func parseResources(result *structs.Resources, list *ast.ObjectList) error {
 			return multierror.Prefix(err, "resources, network, ports ->")
 		}
 
-		result.Networks = []*structs.NetworkResource{&r}
+		result.Networks = []*api.NetworkResource{&r}
 	}
 
 	// Combine the parsed resources with a default resource block.
-	min := structs.DefaultResources()
+	min := &api.Resources{
```
Comment (dadgar) [inline jobspec/parse.go]: Create a min resources for API

File: command/agent/job_endpoint.go (hunk 7 above)
Comment (diptanu) [inline command/agent/job_endpoint.go]: Shouldn't be `PUT`, we should check for `POST`

Comment (diptanu) [inline command/agent/job_endpoint.go]: Well looks like `client.Write` sets the method to `PUT`.

File: command/agent/job_endpoint.go (hunk 6 above)
Comment (diptanu) [inline command/agent/job_endpoint.go]: Default is false, so don't need it.

File: nomad/job_endpoint.go (hunk 16 above)
Comment (diptanu) [inline nomad/job_endpoint.go]: Not relevant since the `ValidationErrors` is a slice of strings.

File: api/jobs.go [hunk 22]
```diff
@@ -237,12 +279,72 @@ type Job struct {
...
+	if j.CreateIndex == nil {
+		j.CreateIndex = helper.Uint64ToPtr(0)
+	}
+	if j.ModifyIndex == nil {
+		j.ModifyIndex = helper.Uint64ToPtr(0)
+	}
+	if j.JobModifyIndex == nil {
+		j.JobModifyIndex = helper.Uint64ToPtr(0)
+	}
+	if j.Periodic != nil {
```
Comment (dadgar) [inline api/jobs.go]: Similar block for update strategy and parameterized?

File: api/tasks.go [hunk 23]
```diff
@@ -157,36 +230,91 @@ type Task struct {
...
+func (tmpl *Template) Canonicalize() {
+	if tmpl.ChangeMode == nil {
+		tmpl.ChangeMode = helper.StringToPtr("restart")
+	}
+	if tmpl.Splay == nil {
+		tmpl.Splay = helper.TimeToPtr(5 * time.Second)
+	}
+	if tmpl.Perms == nil {
+		tmpl.Perms = helper.StringToPtr("0644")
+	}
```
Comment (dadgar) [inline api/tasks.go]: Can you add a case for `if *tmpl.ChangeMode == "signal" && tmpl.ChangeSignal == nil { tmpl.ChangeSignal = helper.StringToPtr("SIGHUP") }`

File: api/tasks.go [hunk 24]
```diff
@@ -157,36 +230,91 @@ type Task struct {
...
+
+func (v *Vault) Canonicalize() {
+	if v.Env == nil {
+		v.Env = helper.BoolToPtr(true)
+	}
+	if v.ChangeMode == nil {
+		v.ChangeMode = helper.StringToPtr("restart")
+	}
+	if v.ChangeSignal == nil {
+		v.ChangeSignal = helper.StringToPtr("sighup")
```
Comment (dadgar) [inline api/tasks.go]: Capitalize the signal both in that helper and put an else if it is set and upper case it

File: api/tasks.go (hunk 23 above)
Comment (dadgar) [inline api/tasks.go]: If it is set upper case the signal

File: api/resources.go [hunk 25]
```diff
@@ -1,15 +1,48 @@
...
+	if other.IOPS != nil {
+		r.IOPS = other.IOPS
+	}
+	if len(other.Networks) != 0 {
+		r.Networks = other.Networks
+	}
+}
+
 type Port struct {
 	Label string
```
Comment (dadgar) [inline api/resources.go]: Make these pointers and give a default value to the mbits (10 seems fine)

File: command/agent/job_endpoint.go [hunk 26]
```diff
@@ -310,3 +354,202 @@ func (s *HTTPServer) jobDispatchRequest(resp http.ResponseWriter, req *http.Requ
...
+		Status:            *job.Status,
+		StatusDescription: *job.StatusDescription,
+		CreateIndex:       *job.CreateIndex,
+		ModifyIndex:       *job.ModifyIndex,
+		JobModifyIndex:    *job.JobModifyIndex,
+	}
+
+	j.Constraints = make([]*structs.Constraint, len(job.Constraints))
+	for i, c := range job.Constraints {
+		j.Constraints[i] = &structs.Constraint{
```
Comment (dadgar) [inline command/agent/job_endpoint.go]: Should be a method so it can be reused

File: api/tasks.go [hunk 27]
```diff
@@ -103,8 +125,43 @@ type TaskGroup struct {
...
+	}
+	if g.EphemeralDisk == nil {
+		g.EphemeralDisk = DefaultEphemeralDisk()
+	} else {
+		g.EphemeralDisk.Canonicalize()
+	}
+	if g.RestartPolicy == nil {
+		switch jobType {
+		case "service", "system":
+			g.RestartPolicy = &RestartPolicy{
```
Comment (dadgar) [inline api/tasks.go]: Make these pointers and add a merge method. Get the default and the user submitted one and merge them

File: command/validate.go [hunk 28]
```diff
@@ -43,17 +43,26 @@ func (c *ValidateCommand) Run(args []string) int {
...
+	}
+
+	// Force the region to be that of the job.
+	if r := job.Region; r != nil {
+		client.SetRegion(*r)
+	}
 
 	// Check that the job is valid
-	if err := job.Validate(); err != nil {
+	if _, _, err := client.Jobs().Validate(job, nil); err != nil {
```
Comment (dadgar) [inline command/validate.go]: Put a warning message in orange if the drivers were not validated

File: nomad/job_endpoint.go [hunk 29]
```diff
@@ -285,6 +285,25 @@ func (j *Job) Summary(args *structs.JobSummaryRequest,
 	return j.srv.blockingRPC(&opts)
 }
 
+// Validate validates a job
+func (j *Job) Validate(args *structs.JobValidateRequest,
```
Comment (dadgar) [inline nomad/job_endpoint.go]: HTTP documentation

File: command/agent/job_endpoint.go [hunk 30]
```diff
@@ -111,6 +113,48 @@ func (s *HTTPServer) jobPlan(resp http.ResponseWriter, req *http.Request,
...
+	args := structs.JobValidateRequest{
+		Job: job,
+		WriteRequest: structs.WriteRequest{
+			Region: validateRequest.Region,
+		},
+	}
+	s.parseRegion(req, &args.Region)
+
+	var out structs.JobValidateResponse
+	if err := s.agent.RPC("Job.Validate", &args, &out); err != nil {
```
Comment (schmichael) [inline command/agent/job_endpoint.go]: Should we do something with this `err` even if no other errs happen?

File: command/agent/job_endpoint.go [hunk 31]
```diff
@@ -111,6 +113,48 @@ func (s *HTTPServer) jobPlan(resp http.ResponseWriter, req *http.Request,
...
+		// Fall back to do local validation
+		args.Job.Canonicalize()
+		if vErr := args.Job.Validate(); vErr != nil {
+			if merr, ok := err.(*multierror.Error); ok {
+				for _, err := range merr.Errors {
+					out.ValidationErrors = append(out.ValidationErrors, err.Error())
+				}
+			}
+		} else {
+			out.ValidationErrors = append(out.ValidationErrors, vErr.Error())
```
Comment (schmichael) [inline command/agent/job_endpoint.go]: Isn't vErr nil here?

File: command/agent/job_endpoint.go [hunk 32]
```diff
@@ -111,6 +113,48 @@ func (s *HTTPServer) jobPlan(resp http.ResponseWriter, req *http.Request,
...
+	}
+	s.parseRegion(req, &args.Region)
+
+	var out structs.JobValidateResponse
+	if err := s.agent.RPC("Job.Validate", &args, &out); err != nil {
+
+		// Fall back to do local validation
+		args.Job.Canonicalize()
+		if vErr := args.Job.Validate(); vErr != nil {
+			if merr, ok := err.(*multierror.Error); ok {
```
Comment (schmichael) [inline command/agent/job_endpoint.go]: Did you mean vErr here?

File: jobspec/parse.go [hunk 33]
```diff
@@ -696,7 +711,11 @@ func parseTasks(jobName string, taskGroupName string, result *[]*structs.Task, l
 		}
 
 		// If we have logs then parse that
-		logConfig := structs.DefaultLogConfig()
+		logConfig := &api.LogConfig{
+			MaxFiles:      helper.IntToPtr(10),
+			MaxFileSizeMB: helper.IntToPtr(10),
+		}
+
```
Comment (schmichael) [inline jobspec/parse.go]: Why not use `api.DefaultLogConfig()`?

**Comment by phinze** [comment]
Hi folks!

It seems this PR changes `nomad validate` to require that an agent be running? Our CI runs `nomad validate` on our job files - do we need to update it to spin up a local agent for itself after this change?

If all this is accurate, I think we should call this out in the "Backwards Incompatibilities" section of the changelog.

**Comment by dadgar** [comment]
@phinze Hey this was not suppose to happen! It was an oversight and 0.5.5 RC2 will restore the original behavior.

**Pull request merged**